runners: Add lldbac runner for run-lldbac debugging

AFOliveira opened the pull request:
This PR introduces a new Zephyr runner `lldbac` that provides interactive debugging support for ARC-V targets using the `run-lldbac` tool from Synopsys ARC MetaWare Development Toolkit (MWDT).

The runner supports both simulator (nSIM) and physical hardware debugging. For simulator mode, it uses `--nsim` with board-configured properties or optional `--tcf` files. For hardware mode, it provides flash and debug capabilities via board.json JTAG configuration. Hardware mode requires explicit `--board-json` argument to prevent accidental misconfigurations.

The `debug` command is available for both simulator and hardware, while `flash` is hardware-only (simulator users should use the existing `arc-nsim` runner for non-interactive execution). TCF and props are mutually exclusive for simulator mode, with TCF taking precedence if both are specified.

Example usage:
```bash
# Simulator debugging
west debug --runner lldbac

# Hardware debugging and flashing
west debug --runner lldbac --hardware --board-json=board.json
west flash --runner lldbac --hardware --board-json=board.json
```

Tested on `nsim_arc_v/rmx100` board target.

AFOliveira committed:
runners: Add lldbac runner for run-lldbac debugging

Introduce a new Zephyr runner, `lldbac`, providing interactive debugging
for both nSIM (simulator) and physical hardware using the run-lldbac
tool from Synopsys ARC MWDT.

This runner provides integrated debugging support:
- debug: Interactive debugging with integrated server management,
  Available for both simulator and hardware (integrated mode)
- flash: Flash and run on physical hardware,
  Available for hardware only (simulator should use arc-nsim)

Key features:
- Hardware mode: Flash and debug on physical hardware using JTAG flags
- Simulator mode: Interactive debugging on nSIM using run-lldbac with
  --nsim flag (props or --tcf). Enable with --simulator flag.
- Board-driven config: Boards can specify defaults in board.cmake
- TCF support: Optional --tcf for simulator mode (mutually exclusive
  with --nsim-props)
- GUI support: Optional --gui flag for VS Code debugging
- Postconnect support: Execute commands after connection via
  --postconnect-cmd and --postconnect-file for hardware initialization
- Board JSON: Optional --board-json for complex board configurations

Hardware configuration (flag-based):
- --jtag-device: JTAG device name (optional, auto-detected)
- --jtag: JTAG adapter type (default: jtag-digilent)
- --jtag-frequency: JTAG clock frequency (default: 500KHz)
- --postconnect-cmd: Command to execute after connection
  (can be used multiple times)
- --postconnect-file: File containing commands to execute after
  hardware initialization
- --board-json: Path to board.json file for board configuration

Simulator configuration:
- --simulator: Enable simulator mode (required)
- --nsim-props: nSIM properties file path
- --tcf: Tool Configuration File (alternative to --nsim-props)

Signed-off-by: Afonso Oliveira <afonsoo@synopsys.com>
Signed-off-by: Mohamed Moawad <moawad@synopsys.com>

AFOliveira committed:
doc: boards: nsim: arc_v: Document lldbac runner usage

Add comprehensive documentation for using the lldbac runner with arc_v
nSIM board targets, covering simulator and hardware debugging workflows.

Signed-off-by: Afonso Oliveira <afonsoo@synopsys.com>
Signed-off-by: Mohamed Moawad <moawad@synopsys.com>

pdgendt commented:
Re-assigning because I have zero experience with ARC

ruuddw reviewed (changes requested):
Good start, but it seems to need some more work/discussion/testing.
E.g. on Windows I had to change the name of the command to "run-lldbac.bat" to ensure the executable (script) was found.

AFOliveira commented on scripts/west_commands/runners/lldbac.py:
```diff
@@ -0,0 +1,278 @@
+
+    def __init__(self, cfg, props, tcf=None, hardware=False, board_json=None, gui=False):
+        super().__init__(cfg)
+        self.lldbac_cmd = ['run-lldbac']
```
Thanks for the tip @marc-hb!

AFOliveira committed:
boards: nsim: arc_v: Add lldbac runner support

Add lldbac runner configuration to arc_v nSIM board family for
simulator debugging with run-lldbac tool.

Configure --nsim-props argument using the same NSIM_PROPS variable as
arc-nsim runner for consistency. This enables the lldbac runner to
use the board-specific nSIM properties for simulator mode debugging.

The lldbac runner supports both simulator and hardware modes, with
simulator mode being enabled via the --simulator flag. For hardware
debugging, users need to provide --board-json configuration separately.

Signed-off-by: Afonso Oliveira <afonsoo@synopsys.com>

M-Moawad commented:
## PR Review Summary

I reviewed the PR and made the following updates based on testing:

- Switched to run-lldbac's native `--core` properties instead of `platform connect` (fixed the empty-core error).
- Added and verified `--postconnect-cmd` and `--postconnect-file` for hardware init after connect.
   - Example: `west flash --runner lldbac --jtag-device=JtagHs2 --postconnect-file postconnect.cmd`
- Added and verified `--board-json` for complex configs.
   - Example: `west flash --runner lldbac --board-json board.json`
- Confirmed `west flash`, `west debug`, and simulator `west debugserver` (client mode) all work.

Notes:
- Please add brief docs for `--board-json` and postconnect usage.
- `--jtag-device` looks optional in practice (run-lldbac auto-detects), but the code enforces it.

Patch:
```patch
diff --git a/scripts/west_commands/runners/lldbac.py b/scripts/west_commands/runners/lldbac.py
index 36f484b3e0f..395d3752997 100644
--- a/scripts/west_commands/runners/lldbac.py
+++ b/scripts/west_commands/runners/lldbac.py
@@ -27,6 +27,9 @@ class LldbacBinaryRunner(ZephyrBinaryRunner):
         gui=False,
         gdb_host=None,
         gdb_port=DEFAULT_LLDBAC_GDB_PORT,
+        postconnect_cmds=None,
+        postconnect_file=None,
+        board_json=None,
     ):
         super().__init__(cfg)
 
@@ -49,6 +52,9 @@ class LldbacBinaryRunner(ZephyrBinaryRunner):
         self.gui = gui
         self.gdb_host = gdb_host
         self.gdb_port = gdb_port
+        self.postconnect_cmds = postconnect_cmds or []
+        self.postconnect_file = postconnect_file
+        self.board_json = board_json
 
     @classmethod
     def name(cls):
@@ -102,6 +108,23 @@ class LldbacBinaryRunner(ZephyrBinaryRunner):
         # Common options
         parser.add_argument('--gui', action='store_true', help='Launch VS Code GUI for debugging')
 
+        # Postconnect options
+        parser.add_argument(
+            '--postconnect-cmd',
+            action='append',
+            help='Command to execute after connection (can be used multiple times)',
+        )
+        parser.add_argument(
+            '--postconnect-file',
+            help='File containing commands to execute after connection',
+        )
+
+        # Board configuration
+        parser.add_argument(
+            '--board-json',
+            help='Path to board.json file for board configuration',
+        )
+
     @classmethod
     def do_create(cls, cfg, args):
         return LldbacBinaryRunner(
@@ -115,6 +138,9 @@ class LldbacBinaryRunner(ZephyrBinaryRunner):
             gui=args.gui,
             gdb_host=args.gdb_host,
             gdb_port=args.gdb_port,
+            postconnect_cmds=args.postconnect_cmd,
+            postconnect_file=args.postconnect_file,
+            board_json=args.board_json,
         )
 
     def do_run(self, command, **kwargs):
@@ -138,28 +164,34 @@ class LldbacBinaryRunner(ZephyrBinaryRunner):
                 "'west flash --runner arc-nsim'"
             )
 
-        if not self.jtag_device:
-            raise ValueError(
-                "Hardware flash requires --jtag-device. Example:\n"
-                "  west flash --runner lldbac --jtag-device=JtagHs2"
-            )
-
         # Require run-lldbac for flash
         self.require(self.run_lldbac_cmd)
 
-        # Build platform connect command
-        connect_cmd = self._build_platform_connect_cmd()
-
-        # Create run-lldbac script for hardware flash
+        # Create run-lldbac command for hardware flash
         lldbac_cmd = [
             self.run_lldbac_cmd,
             '--batch',  # Exit after executing commands
-            '-o',
-            connect_cmd,
-            '-o',
-            'run',
         ]
 
+        # Use board.json if provided, otherwise build core properties
+        if self.board_json:
+            lldbac_cmd.extend(['--cores-json', self.board_json])
+            self.logger.info(f"Using board configuration from: {self.board_json}")
+        else:
+            if not self.jtag_device:
+                raise ValueError(
+                    "Hardware flash requires --jtag-device. Example:\n"
+                    "  west flash --runner lldbac --jtag-device=JtagHs2"
+                )
+            # Build core properties for hardware connection
+            core_props = self._build_core_properties_for_hardware()
+            lldbac_cmd.extend(['--core', core_props])
+
+        # Add postconnect commands
+        lldbac_cmd.extend(self._build_postconnect_args())
+
+        lldbac_cmd.extend(['-o', 'run'])
+
         if self.gui:
             lldbac_cmd.append('--gui')
 
@@ -175,7 +207,11 @@ class LldbacBinaryRunner(ZephyrBinaryRunner):
 
         lldbac_cmd = [self.run_lldbac_cmd]
 
-        if self.simulator:
+        # Use board.json if provided (works for both simulator and hardware)
+        if self.board_json:
+            lldbac_cmd.extend(['--cores-json', self.board_json])
+            self.logger.info(f"Using board configuration from: {self.board_json}")
+        elif self.simulator:
             # Simulator mode: use --nsim or --tcf
             if self.tcf:
                 lldbac_cmd.extend(['--tcf', self.tcf])
@@ -186,20 +222,23 @@ class LldbacBinaryRunner(ZephyrBinaryRunner):
                 lldbac_cmd.extend(['--nsim', nsim_props_string])
                 self.logger.info(f"Starting integrated simulator debug (props: {nsim_props_file})")
         else:
-            # Hardware mode: use platform connect
+            # Hardware mode: use --core with connection properties
             if not self.jtag_device:
                 raise ValueError(
                     "Hardware debug requires --jtag-device. Example:\n"
                     "  west debug --runner lldbac --jtag-device=JtagHs2"
                 )
 
-            connect_cmd = self._build_platform_connect_cmd()
-            lldbac_cmd.extend(['-o', connect_cmd])
+            core_props = self._build_core_properties_for_hardware()
+            lldbac_cmd.extend(['--core', core_props])
             self.logger.info(
                 f"Starting integrated hardware debug "
                 f"(JTAG: {self.jtag}, device: {self.jtag_device})"
             )
 
+        # Add postconnect commands
+        lldbac_cmd.extend(self._build_postconnect_args())
+
         # Load ELF in debug mode
         lldbac_cmd.extend(['-o', 'load'])
 
@@ -241,17 +280,38 @@ class LldbacBinaryRunner(ZephyrBinaryRunner):
         self.logger.debug(f"Running command: {' '.join(lldbac_cmd)}")
         self.check_call(lldbac_cmd)
 
-    def _build_platform_connect_cmd(self):
-        '''Build platform connect command from JTAG flags.'''
-        parts = [f'platform connect --connect={self.jtag}']
+    def _build_core_properties_for_hardware(self):
+        '''Build core properties string for run-lldbac --core argument.
+
+        Format: "connect=<type> connect-props=key1=val1 connect-props=key2=val2"
+        '''
+        props = [f'connect={self.jtag}']
 
         if self.jtag_device:
-            parts.append(f'--jtag-device={self.jtag_device}')
+            props.append(f'connect-props=jtag_device={self.jtag_device}')
 
         if self.jtag_frequency:
-            parts.append(f'--jtag-frequency={self.jtag_frequency}')
+            props.append(f'connect-props=jtag_frequency={self.jtag_frequency}')
+
+        return ' '.join(props)
+
+    def _build_postconnect_args(self):
+        '''Build postconnect arguments for run-lldbac.
+
+        Returns a list of arguments to add to the run-lldbac command.
+        '''
+        args = []
+
+        # Add postconnect commands from --postconnect-cmd
+        if self.postconnect_cmds:
+            for cmd in self.postconnect_cmds:
+                args.extend(['--postconnect', cmd])
+
+        # Add postconnect file from --postconnect-file
+        if self.postconnect_file:
+            args.extend(['--postconnect', f'command source {self.postconnect_file}'])
 
-        return ' '.join(parts)
+        return args
 
     def _get_nsim_props(self):
         '''Get nSIM properties file path: board_dir/support/<props>.'''


```

marc-hb commented:
> shutil.which() automatically appends .bat or .exe on Windows, which solves the Windows problem reported by @ruuddw in https://github.com/zephyrproject-rtos/zephyr/pull/97635#pullrequestreview-3344455129
> This is already used in dozens of places in this repo.

I just discovered that this technique is not just a "trick" but actually part of the official Python documentation: https://docs.python.org/3/library/subprocess.html#popen-constructor. 
Good to know!

> To search for an unqualified name on PATH, use [shutil.which()](https://docs.python.org/3/library/shutil.html#shutil.which).

ruuddw reviewed (changes requested):
Please don't use "debugserver" to connect to a server, it should launch the server.

pdgendt commented on scripts/west_commands/runners/lldbac.py:
```diff
@@ -0,0 +1,290 @@
+        )
+
+        # Common options
+        parser.add_argument('--gui', action='store_true', help='Launch VS Code GUI for debugging')
```
Ditto

```suggestion
        parser.add_argument(
            '--gui',
            action=argparse.BooleanOptionalAction,
            help='Launch VS Code GUI for debugging'
        )
```

M-Moawad commented on scripts/west_commands/runners/lldbac.py:
```diff
@@ -0,0 +1,290 @@
+            '--simulator',
+            action='store_true',
+            help='Use simulator instead of physical hardware (integrated mode)',
+        )
```
Thanks for the suggestion, updated.
